src/error.rs

```diff
@@ -0,0 +1,222 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! Error code for the `hex-conservative` crate.
+
+use core::convert::Infallible;
+use core::fmt;
+
+/// Formats error.
+///
+/// If `std` feature is OFF appends error source (delimited by `: `). We do this because
+/// `e.source()` is only available in std builds, without this macro the error source is lost for
+/// no-std builds.
+macro_rules! write_err {
+    ($writer:expr, $string:literal $(, $args:expr)*; $source:expr) => {
+        {
+            #[cfg(feature = "std")]
+            {
+                let _ = &$source;   // Prevents clippy warnings.
+                write!($writer, $string $(, $args)*)
+            }
+            #[cfg(not(feature = "std"))]
+            {
+                write!($writer, concat!($string, ": {}") $(, $args)*, $source)
+            }
+        }
+    }
+}
+pub(crate) use write_err;
+
+/// Hex decoding error while parsing to a vector of bytes.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum DecodeToBytesError {
```

Comment:
Sorry for poking into naming again but I just realized this name is inconsistent with `DecodeToArrayError` since this one doesn't refer to a type but `Array` does. And I also realized mentioning the type leaks internal details about functions that call our functions.

So I think we should refer to the property that distinguishes them: length being fixed or dynamic. Thus `DecodeDynSizedError` and `DecodeFixedSizedError`.

But if you're really annoyed by this bikeshedding, I can accept "shut up and get it released". I kinda suspect the downstream errors will want to map the enum to something else instead so "leaking internals" is not going to happen anyway.